packages/rum/package.json

```diff
@@ -26,9 +26,9 @@
     "test:node": "npm run script runNodeTests",
     "test:integration": "npm run script runIntegrationTests",
     "test:unit": "node ../../dev-utils/run-script.js runUnitTests false packages/rum",
-    "test:e2e:supported": "npm run script runE2eTests wdio.conf.js",
-    "test:e2e:failsafe": "npm run script runE2eTests wdio-failsafe.conf.js",
-    "test:sauce": "npm run script runSauceTests",
+    "test:e2e:supported": "node ../../dev-utils/run-script.js runE2eTests packages/rum/wdio.conf.js",
+    "test:e2e:failsafe": "node ../../dev-utils/run-script.js runE2eTests packages/rum/wdio-failsafe.conf.js",
+    "test:sauce": "node ../../dev-utils/run-script.js runSauceTests true packages/rum test:unit test:e2e:supported test:e2e:failsafe",
```

Comment:
this seems wrong. the point of sauce script tests was to run unit and e2e in parallel.

Reply:
```js
await runAll('test:unit', loggerOpts)
      await runAll(['test:e2e:supported', 'test:e2e:failsafe'], {
        parallel: true,
        printLabel: true,
        ...loggerOpts
      })
```

The set up should look similar to this, Right now it is not.

Reply:
Since this method is generic we need to provide parallel tasks separately and I think that makes script more complicated. Furthermore, I think it's better to increase the parallel factor of individual tasks instead of running two task at the same time.

Reply:
Not really, its better to at least run the e2e tests parallel. Plus the script takes so many args and its better to just separate those instead of making it more generic. 

```js

// Not exported
function runSauceTests(takstoRun, flag) {
 // stays the same
// Decide on serial parallel here based on flag
}

/ / Exported method
export function startSauceTest(packageName) {
  const tasks = []
  
  if(packageName === "rum"  {
    tasks.push(...)
  } else if(packageName === "react"){
      // do something here
  }

 return runSauceTest(tasks, flag)
}
```

Package.json would look like
```json
"test:sauce":  "node ./run-script startSauceTest package/rum"
```

Currently both the `package.json` and script seems to everything together with lots of parameters. Its better to move them to a single place. Makes it easier to understand and also reason about.

Reply:
I think the difference in performance is not significant, if any at all. 
The approach in your code is not generic, it has special cases for packages. this is not scalable for multiple packages and it's hard to keep track of.

IMO, it's better to consume all possible sessions for each individual tasks. why do you think running two tasks in parallel has any advantage?

Reply:
Also running in parallel makes finding failures and keeping track of tests harder since the failsafe and supported results are mixed together:

![image](https://user-images.githubusercontent.com/6134122/59438808-7a020480-8df4-11e9-9f49-9b7893ccb48a.png)

Reply:
> I think the difference in performance is not significant, if any at all.

Do we have any numbers here? 

> The approach in your code is not generic, it has special cases for packages. this is not scalable for multiple packages and it's hard to keep track of.

No ,ignore the tasks on whether its serial or parallel. Look at the number of flags that is getting passed to the package.json script. It looks totally weird. I proposed to have them in a single place inside single function. 

```json
"test:sauce":  "node ./run-script startSauceTest package/rum"
```

> Also running in parallel makes finding failures and keeping track of tests harder since the failsafe and supported results are mixed together:

its just output aggregation, you can just set `aggregateOutput`: true and it would fix it.

Reply:
I am fine with keeping tasks in serial if tests time delta is not significant. But lets try to keep the package.json a bit clean. It really gets messy by keeping list of tasks there along with other args.

Reply:
I don't see an issue with the number of arguments send to a function, the issue I think is that they don't have names so ideally it should be `node ./run-script startSauceTest --package=package/rum`. similar to how we use lerna `lerna run test --stream --scope=$SCOPE --no-prefix`

In any case, we can always iterate on this, I don't think we should fix everything in this PR.

Reply:
We have been doing this iteration for quite some time and that was the reason why I suggested this change. What I meant is to keep the package.json clean and scripts convey what its doing instead of making everything generic. 
```json
"test: sauce": "node ./run-script startSauceTest --package=package/rum"
```
will stay the same and just the tasks will be moved closer to the script. 

```js
const TASKS_FOR_RUM = ["unit", "e2e"]
function startSauceTest(package) {
   let tasks = []
    if(package==="rum") . {
        tasks.concat(TASKS_FOR_RUM)
   }
}
```

I am fine with merging this PR as it as now.

Reply:
FWIW, I think having special cases in a generic script is not a good idea and certainly makes the script harder to follow when we have 3 or more packages, but I'm open to other suggestions.

Thanks for the review 👍

Reply:
My whole reasoning behind it was bloated `package.json`. If we have to run another taks, the list keeps on growing. Its better to keep them in a single place where generic script lies. We can merge the PR and revisit it later if needed.  Lets not spend more time on this 😄